beyond-all-reason/spring
# Fallback fonts with atlas clearing.

### Work done

- Introduce lua `AddFallbackFont(vfspath)` and `ClearFallbackFonts()`.
  - Backend at CFontTexture::AddFallbackFont and CFontTexture::ClearFallbackFonts.
- Add the mechanism so atlases are properly cleared
- Adds the `FontsChanged` callin to signal applications text dlists or other text caches need to be refreshed.

### Remarks

- To review, compare directly, or to "fallback fonts" branch: **[direct link to comparison](https://github.com/saurtron/spring/compare/lua-fallback-fonts...saurtron:spring:lua-fallback-fonts-with-atlasclear)**  to better see the clearing code 
- Added it as separate PR to facilitate review.
  - ~~Added it as DRAFT to avoid accidental merging before [#1777]( https://github.com/beyond-all-reason/spring/pull/1777). It can be merged at the same time by just merging this PR, or can be merged right after that one.~~
  - This is now the "fallback-fonts" branch I propose for inclusion, we can still merge only [#1777]( https://github.com/beyond-all-reason/spring/pull/1777) but otherwise that one will serve just as reference from now on.
  - - Compared to #1777, this adds atlas clearing, so adding and clearing fallbacks will be properly reflected by the game at any time.
- Clearing atlases might not be critical, since most applications are expected to define glyphs at the beginning and not change too much, but it's needed for robustness.

### New api

#### gl.AddFallbackFont(vfspath)

Add a fallback font. First added will have higher priority.

Link to [code](https://github.com/beyond-all-reason/spring/pull/1790/files#diff-57a581feb9995582f88afe5c097dc7b08376e57b735fb3e8871017f69938c23cR227)  and inline documentation.

#### gl.ClearFallbackFonts()

Clear all fallback fonts.

Link to [code](https://github.com/beyond-all-reason/spring/pull/1790/files#diff-57a581feb9995582f88afe5c097dc7b08376e57b735fb3e8871017f69938c23cR246) and inline documentation.

#### unsynced callin 'FontsChanged'.

This is needed so the game/application can properly react to texture atlases having been changed. Modules usually cook font drawing into display lists, and those need to be cleared when receiving this callin, otherwise previously cooked text fallbacks won't show properly (base glyphs will, since we restore them into the same atlas coordinates).

### Testing

* Get this bar branch [test-fallback-fonts](https://github.com/saurtron/Beyond-All-Reason/tree/test-fallback-fonts) into your BAR.sdd
   * This branch has functionality for enabling/disabling one fallback through mouse left click, and also runs an automatic procedure to write into chat and enable/disable fallbacks.
   * Also listens for FontsChanged callin in order to clear dlists at gui_chat.
* Open skirmish
* Once it starts, the test widget will automatically run the following procedure:
  * Write something into chat, like: "blah🔥" (at frame 3)
    * it will use your os fallback
  * Enable fallbacks and write into chat (at frame 150)
    * should say "Fallback fonts, true"
    * The fire glyph should change to the provided font
  * Disable fallbacks and write into chat (at frame 450)
    * should say "Fallback fonts, false" 
    * The fire glyph should change to the os font again


You can further play with it by using left mouse button to enable/disable fallback, and writing your own characters into chat.

### After

writing with fallback for Noto enabled, then disabling it (note how the fire symbol reverts to the system default), then adding the fallback again. (this uses the new callin to refresh dlists at gui_chat, otherwise it wouldn't refresh properly).

![refreshatlas](https://github.com/user-attachments/assets/5d1cccfe-6d1a-4a29-94c6-5460cc6647a5)
